Add new metric to show duration all pipelineruns have taken

zakisk opened the pull request:
added new metric to show sum of durations all pipelineruns have taken in seconds. added docs and adjusted test accordingly

https://issues.redhat.com/browse/SRVKP-6226

![Screenshot from 2024-09-20 16-55-20](https://github.com/user-attachments/assets/998f2476-d625-4fc3-a7fe-46e11d4b7063)


# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 Please ensure your commit message is clear and informative. For guidance on crafting effective commit messages, refer to the How to write a git commit message guide. We prefer the commit message to be included in the PR body itself rather than a link to an external website (ie: Jira ticket).

- [ ] ♽ Before submitting a PR, run make test lint to avoid unnecessary CI processing. For an even more efficient workflow, consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the root of this repository.

- [ ] ✨ We use linters to maintain clean and consistent code. Please ensure you've run make lint before submitting a PR. Some linters offer a --fix mode, which can be executed with the command make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) tools are installed first).

- [ ] 📖 If you're introducing a user-facing feature or changing existing behavior, please ensure it's properly documented.

- [ ] 🧪 While 100% coverage isn't a requirement, we encourage unit tests for any code changes where possible.

- [ ] 🎁 If feasible, please check if an end-to-end test can be added. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.

- [ ] 🔎 If there's any flakiness in the CI tests, don't necessarily ignore it. It's better to address the issue before merging, or provide a valid reason to bypass it if fixing isn't possible (e.g., token rate limitations).

zakisk commented:
/test

zakisk commented:
/test

chmouel commented on pkg/reconciler/emit_metrics_test.go:
```diff
@@ -2,12 +2,16 @@ package reconciler
+	duckv1 "knative.dev/pkg/apis/duck/v1"
 )
 
 func TestEmitMetrics(t *testing.T) {
```
you are not adding any new tests here, just fixing the current one?

zakisk replied:
yeah, adjusted existing one to accommodate new metric.

chmouel replied:
can we have a new test?

zakisk replied:
can't figure out how, because there is only one func `emit_metrics`, can you please guide how?

chmouel replied:
I do not know how, without having to spend time on the code, what about a e2e test then?

zakisk replied:
something like done in pipeline [metric tests](https://github.com/tektoncd/pipeline/blob/main/pkg/pipelinerunmetrics/metrics_test.go)?

chmouel commented on pkg/metrics/metrics.go:
```diff
@@ -96,3 +120,25 @@ func (r *Recorder) Count(provider, event, namespace, repository string) error {
+func (r *Recorder) CountPRDuration(namespace, repository, status, reason string, duration time.Duration) error {
+	if !r.initialized {
+		return fmt.Errorf(
+			"ignoring the metrics recording for pipeline runs,  failed to initialize the metrics recorder")
```
pipelineruns and remove the double spacing and i think this can use a rewording

chmouel commented on docs/content/docs/install/metrics.md:
```diff
@@ -10,6 +10,7 @@ The metrics for pipelines-as-code can be accessed through the `pipelines-as-code
+| Name                                                 | Type    | Description                                                        |
+|------------------------------------------------------|---------|--------------------------------------------------------------------|
+| `pipelines_as_code_pipelinerun_count`                | Counter | Number of pipelineruns created by pipelines-as-code                |
+| `pipelines_as_code_pipelinerun_duration_seconds_sum` | Counter | Number of seconds all pipelineruns have taken in pipelines-as-code |
```
is it the sum taken of all pipelinerun install?

what is the usefulness of this ? was this coming from a customer/user request?

zakisk replied:
@chmouel console team wants total number of pipelineruns metric (which we already have) and sum of all duration so that they can show average time taken by pipelinerun in specified time range.

slack thread: https://redhat-internal.slack.com/archives/CG5GV6CJD/p1726742648504399

savitaashture commented on pkg/metrics/metrics.go:
```diff
@@ -15,13 +15,19 @@ var prCount = stats.Float64("pipelines_as_code_pipelinerun_count",
 	stats.UnitDimensionless)
 
+var prDurationCount = stats.Float64("pipelines_as_code_pipelinerun_duration_seconds_sum",
+	"number of second pipelineruns completed in by pipelines as code",
```
```suggestion
	"number of seconds pipelineruns completed in by pipelines as code",
```

savitaashture commented on pkg/metrics/metrics.go:
```diff
@@ -59,13 +65,31 @@ func NewRecorder() (*Recorder, error) {
+			Description: prDurationCount.Description(),
+			Measure:     prDurationCount,
+			Aggregation: view.Sum(),
+			TagKeys:     []tag.Key{r.namespace, r.repository, r.status, r.reason},
```
Should we add pipelinerun name also which give info of which pipelinerun has max duration :thinking: 

as repo can have multiple pipelinerun associated

zakisk replied:
@savitaashture console team only wants sum of duration taken by all pipelineruns, I asked about pipelinerun name tag and it isn't wanted by team, and also it will affect on cardinality of  metric if we put pipelinerun tag.

savitaashture replied:
I see okay
then its fine to not add

savitaashture commented on pkg/reconciler/emit_metrics.go:
```diff
@@ -25,5 +28,43 @@ func (r *Reconciler) emitMetrics(pr *tektonv1.PipelineRun) error {
+	if pr.Status.StartTime != nil {
+		duration = time.Since(pr.Status.StartTime.Time)
+		if pr.Status.CompletionTime != nil {
+			duration = pr.Status.CompletionTime.Sub(pr.Status.StartTime.Time)
```
Here i m trying to understand 
if pr.Status.CompletionTime is nil then displayed duration  is just startTime initially and is it reconciling again to update the duration when PipelineRun completes :thinking:

zakisk replied:
`emit_metrics` is called in [reportFinalStatus](https://github.com/openshift-pipelines/pipelines-as-code/blob/b3fd3ef866c6ddb72cd1dee195f82e169a89a393/pkg/reconciler/reconciler.go#L196) func which is called only once for a pipelinerun (on pipelinerun is completed).

savitaashture commented on pkg/reconciler/emit_metrics.go:
```diff
@@ -25,5 +28,43 @@ func (r *Reconciler) emitMetrics(pr *tektonv1.PipelineRun) error {
+	err := r.metrics.CountPRDuration(pr.GetNamespace(), repository, status, reason, duration)
+	if err != nil {
+		return err
+	}
```
```suggestion
	if err := r.metrics.CountPRDuration(pr.GetNamespace(), repository, status, reason, duration); err != nil {
		return err
	}
```

savitaashture reviewed (commented):
Thank you @zakisk 
Few comments and  also can you add e2e to test both the metrics

zakisk committed:
Add new metric to show duration all pipelineruns have taken

added new metric to show sum of durations all pipelineruns have
taken in seconds. added docs and test accordingly

https://issues.redhat.com/browse/SRVKP-6226

Signed-off-by: Zaki Shaikh <zashaikh@redhat.com>

zakisk commented:
/test

zakisk commented:
> Thank you @zakisk Few comments and also can you add e2e to test both the metrics

@savitaashture about adding E2E tests, I will need to do some extra work, like adding an ingress to access metrics endpoint in tests, metrics endpoint doesn't return response in JSON so need to figure out a consistent way to find metrics values in response so that it will be reusable across all PAC metrics E2E tests, so I am thinking doing that in a separate SRVKP.
CC: @chmouel

savitaashture commented:
/lgtm

Thank you
